feat(http): export a list of http status codes

QuentinFchx opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/angular/angular/issues/23543


## What is the new behavior?
The `@angular/common/http` now exports a list of http status codes.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
I would need directions if this needs to be added to the documentation.

AhnpGit commented on packages/common/http/src/status_codes.ts:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Http status codes.
+ */
+export enum HttpStatusCode {
```
I think it should be const enum like the discussion in #23543

QuentinFchx replied:
Yup, my bad. I wanted to be consistent with existing enums declaration (such as [ChangeDetectionStrategy](https://github.com/angular/angular/blob/master/packages/core/src/change_detection/constants.ts) ).

alfaproject commented:
If you export as const then it won’t be in the bundle and can’t be used outside this repo anyway, right?

QuentinFchx commented:
@alfaproject this it what I thought at first, but now I'm a little confused haha 😄.

QuentinFchx commented:
Hello :)
Would it be possible to have a status on this?
If you think the PR is still relevant, I'll fix conflicts and hopefully have this merged 😉

QuentinFchx commented:
I signed the CLA with my new email address!

QuentinFchx commented:
Better late than never 😀

QuentinFchx commented:
Hello @ahnpnl !
Would you have some time to review this by any chance? :bowtie:

ahnpnl commented:
Hi all LGTM, just wonder if 103 is counted because I saw it on MDN.

petebacondarwin commented:
@QuentinFchx - sorry that this has been around for so long. Could you rebase and then I will begin the review process.

mary-poppins commented:
You can preview f070b77 at https://pr23548-f070b77.ngbuilds.io/.
You can preview c014a74 at https://pr23548-c014a74.ngbuilds.io/.
You can preview abb373f at https://pr23548-abb373f.ngbuilds.io/.
You can preview 26dad83 at https://pr23548-26dad83.ngbuilds.io/.

QuentinFchx commented:
No worries, I can do that.

Just to avoid some back and forth, in the current state of the PR I replaced all the "hardcoded" status codes in the tests by their equivalent in the enum.
Is it something we want to keep? It probably means we want some rules to avoid contributors from hardcoding them in the future (when writing new tests) and end up with 2 ways of having status code represented in the code.
However, I'm not sure I'll have the time to dig into that (how to implement such rules).

I'll do two separate commits:

- 1 with the enum + export + some documentation where relevant
- 1 with the status code replaced in the tests

and I'll let you decide if you want to keep both ;p

petebacondarwin commented:
Two commits sounds good. Thanks

QuentinFchx committed:
feat(http): expose a list of human-readable http status codes

They aim to improve code readability.
Since they are defined by `const enum` they have zero runtime performance impact
over just using constant literals.

Fixes #23543

QuentinFchx commented:
Here we go, I pushed the rebase.

I noticed there was already status codes within that repository [here](https://github.com/angular/angular/blob/master/packages/misc/angular-in-memory-web-api/src/http-status-codes.ts) (that were added since I opened the PR).

Also some hard-coded statuses can be found in `/aio` and `packages/service-worker`. I did not update these paths.

mary-poppins commented:
You can preview 5e70f7f at https://pr23548-5e70f7f.ngbuilds.io/.

petebacondarwin commented:
I'm afraid you will need to make friends with our lint check job and make your commit messages a bit longer :-)
Also I think the second commit is actually `test(http): ...`

petebacondarwin commented:
For reference: https://pr23548-5e70f7f.ngbuilds.io/api/common/http/HttpStatusCode

petebacondarwin reviewed (approved):
LGTM - note that since these are defined by `const enum` they have zero runtime performance impact over just using constant literals.

Reviewed-for: public-api

petebacondarwin commented:
Oh and you need to update the golden public API files by running:

```sh
yarn bazel run //packages/common:common_api.accept
```

alxhub commented on packages/common/http/src/response.ts:
```diff
@@ -348,3 +348,78 @@ export class HttpErrorResponse extends HttpResponseBase implements Error {
+  UnsupportedMediaType = 415,
+  RangeNotSatisfiable = 416,
+  ExpectationFailed = 417,
+  ImATeapot = 418,
```
This is the only one here that really matters.

QuentinFchx committed:
test(http): use HttpStatusCode enum in tests

Now that we have introduced human-readable http status codes,
we can use it in our tests to improve clarity.

mary-poppins commented:
You can preview 867187d at https://pr23548-867187d.ngbuilds.io/.

mhevery reviewed (approved):
reviewed-for: global-approvers

mhevery commented:
[presubmit](http://test/OCL:354201370:BASE:354201390:1611794616632:a77301b6)
